### ISSUE TITLE
Local FileExistsErrors when using Cloudpathlib with moto

### ISSUE THREAD

**Issue by medley56** [drivendataorg/cloudpathlib#325]
```
Python 3.11.1
cloudpathlib                    0.13.0
moto                            4.1.4
```

When Cloudpathlib writes a file to S3, it creates a local file buffer to upload (`cloudpath.py:471: in open`). When using moto in a pytest session, it clears out the mocked S3 buckets at the end of each test. Obviously when cleaning up its buckets, moto does not clear out the local file buffer written by Cloudpathlib (`s3path.fspath`). This appears to cause FileExistsErrors when using Cloudpathlib's exclusive write mode functionality.

Example: Two tests that do exactly the same thing -- write a file using `mode='x'` to a path in a mocked S3 bucket. The second test to run raises a FileExistsError due to the exclusive write mode being applied to the local file buffer which wasn't cleared up by moto after the last test. 

Note that `s3path.exists()` returns `False` immediately before we try to write to the bucket.

```
import string
import random
import pytest
import boto3
from moto import mock_s3
from pathlib import Path
from cloudpathlib import S3Path

FILE_CONTENTS = "Lorem ipsum"


@pytest.fixture
def mock_s3_context():
    """Everything under/inherited by this runs in the mock_s3 context manager"""
    with mock_s3():
        # Yield the (mocked) s3 Resource object
        # (see boto3 docs: https://boto3.amazonaws.com/v1/documentation/api/latest/guide/resources.html)
        yield boto3.resource('s3')


@pytest.fixture
def create_mock_bucket(mock_s3_context):
    """Returns a function that allows dynamic creation of s3 buckets with option to specify the name.

    Note: if the bucket already exists, this doesn't overwrite it. Previous contents will remain.
    Caution: If you create multiple objects at the same location, you may get conflicts"""
    s3 = mock_s3_context

    # The following call to Random() creates a locally seeded random generator. This prevents the pytest-randomly
    # seeded global PRN generator from creating the same "random" bucket names for every test.
    local_random = random.Random()

    def _create_bucket(bucket_name: str = None) -> s3.Bucket:
        """Creates a mock bucket, optionally with a custom name.

        Returns
        -------
        : s3.Bucket
        """
        if not bucket_name:
            bucket_name = ''.join(local_random.choice(string.ascii_letters) for _ in range(16))
        bucket = s3.Bucket(bucket_name)
        if not bucket.creation_date:  # If bucket doesn't already exist
            bucket.create()
            print(f"Created mock S3 bucket {bucket}.")
        else:
            print(f"Using existing mock S3 bucket {bucket}.")
        return bucket
    yield _create_bucket


def test_a(create_mock_bucket):
    """Test creating a mock bucket and putting something in it and then retrieving it"""
    bucket = create_mock_bucket('bucket1')
    uri = f"s3://{bucket.name}/test_text.txt"
    s3path = S3Path(uri)
    print(f"Before writing file, s3path.exists() gives {s3path.exists()}")
    if Path(s3path.fspath).exists():
        print(f"The mocked filesystem path {s3path.fspath} already exists")
    with S3Path(uri).open(mode='x') as xwriter:
        xwriter.write(FILE_CONTENTS)
    print(f"After writing file, s3path.exists() gives {s3path.exists()}")
    with s3path.open() as s3txt:
        print(s3txt.read())


def test_b(create_mock_bucket):
    """Test creating a mock bucket and putting something in it and then retrieving it"""
    bucket = create_mock_bucket('bucket1')
    uri = f"s3://{bucket.name}/test_text.txt"
    s3path = S3Path(uri)
    print(f"Before writing file, s3path.exists() gives {s3path.exists()}")
    if Path(s3path.fspath).exists():
        print(f"The mocked filesystem path {s3path.fspath} already exists")
    with S3Path(uri).open(mode='x') as xwriter:
        xwriter.write(FILE_CONTENTS)
    print(f"After writing file, s3path.exists() gives {s3path.exists()}")
    with s3path.open() as s3txt:
        print(s3txt.read())
```

Perhaps a fix for this could be to clean up the local file buffers used by Cloudpathlib?

**Comment by pjbull** [comment]
Thanks for filing the request @medley56, did you have a chance to look at the docs for testing with cloudpathlib?
https://cloudpathlib.drivendata.org/stable/testing_mocked_cloudpathlib/

We abstract all of the backend, so you shouldn't need to use `moto` at all to test code that uses cloudpathlib. If that doesn't work for you for your use case, it would be good to know why.

The other option that is helpful is to look at the documentation for clearing cache files. One of those options may work well with the `moto` approach you have:
https://cloudpathlib.drivendata.org/stable/caching/#clearing-the-file-cache

I'm going to close this for now, but definitely re-open if there is an explicit bug you encounter!

**Comment by medley56** [comment]
Thanks for the response! I had not seen those docs, are they new? I'm pretty sure I looked for a best practice answer to this about a year ago when I first implemented these tests. Regardless, I will read through those and see if they can help me.

That said, we have a somewhat inconsistent use of cloudpathlib throughout our codebase. We started our project and then discovered cloudpathlib part way through so some calls to S3 are still done using boto3 directly and it's convenient to simply mock out the S3 context using moto; we are slowly working on re-implementing those areas to use cloudpathlib but it's honestly not the highest priority.

I also appreciate the note about pytest scope. Nearly all of our fixtures are default (function) scoped and moto tears down the mock buckets between each test but we had a few tests hanging out that were referencing the same bucket names and object paths, thereby generating the same cloudpathlib local cache filename and resulting in the conflict.

With that in mind, I think your suggestion to clear the local file cache is exactly what I need. I'm going to add that to the pytest teardown phase of my fixtures that use mocked S3 buckets. Our tests might slow down a tiny bit but that's fine. They're just tests.

Your library is great and you are great! Thanks for all your hard work on this (now ubiquitous) tool!
